cpp/src/db/db_object.cu

```diff
@@ -0,0 +1,328 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cugraph.h>
+#include <rmm_utils.h>
+#include <db/db_object.cuh>
+#include <cub/device/device_run_length_encode.cuh>
+
+namespace cugraph {
+  // Define kernel for copying run length encoded values into offset slots.
+  template<typename T>
+  __global__ void offsetsKernel(T runCounts, T* unique, T* counts, T* offsets) {
+    uint64_t tid = threadIdx.x + blockIdx.x * blockDim.x;
+    if (tid < runCounts)
+      offsets[unique[tid]] = counts[tid];
+  }
+
+  template<typename idx_t>
+  db_pattern_entry<idx_t>::db_pattern_entry(std::string variable) {
+    is_var = true;
+    variableName = variable;
+  }
+
+  template<typename idx_t>
+  db_pattern_entry<idx_t>::db_pattern_entry(idx_t constant) {
+    is_var = false;
+    constantValue = constant;
+  }
+
+  template<typename idx_t>
+  db_pattern_entry<idx_t>::db_pattern_entry(const db_pattern_entry<idx_t>& other) {
+    is_var = other.is_var;
+    constantValue = other.constantValue;
+    variableName = other.variableName;
+  }
+
+  template<typename idx_t>
+  bool db_pattern_entry<idx_t>::isVariable() const {
+    return is_var;
+  }
+
+  template<typename idx_t>
+  idx_t db_pattern_entry<idx_t>::getConstant() const {
+    return constantValue;
+  }
+
+  template<typename idx_t>
+  std::string db_pattern_entry<idx_t>::getVariable() const {
+    return variableName;
+  }
+
+  template class db_pattern_entry<int32_t> ;
+  template class db_pattern_entry<int64_t> ;
+
+  template<typename idx_t>
+  db_pattern<idx_t>::db_pattern() {
+
+  }
+
+  template<typename idx_t>
+  db_pattern<idx_t>::db_pattern(const db_pattern<idx_t>& other) {
+    for (size_t i = 0; i < other.entries.size(); i++) {
+      entries.push_back(other.getEntry(i));
+    }
+  }
+
+  template<typename idx_t>
+  int db_pattern<idx_t>::getSize() const {
+    return entries.size();
+  }
+
+  template<typename idx_t>
+  const db_pattern_entry<idx_t>& db_pattern<idx_t>::getEntry(int position) const {
+    return entries[position];
+  }
+
+  template<typename idx_t>
+  void db_pattern<idx_t>::addEntry(db_pattern_entry<idx_t>& entry) {
+    entries.push_back(entry);
+  }
+
+  template<typename idx_t>
+  bool db_pattern<idx_t>::isAllConstants() {
+    bool is_all_constants = true;
+    for (size_t i = 0; i < entries.size(); i++)
+      if (entries[i].isVariable())
+        is_all_constants = false;
```

Comment:
you can just return false at this point without finishing the loop.

Reply:
Will make the change.